src/fun_lti_provider/launch_params.py

```diff
@@ -0,0 +1,219 @@
+
+        for param in LAUNCH_PARAMS_REQUIRED:
+            if param not in self:
+                raise MissingLaunchParamException(param)
+
+    def _param_value(self, param: str) -> Union[str, List]:
+        """Get the value of a LTI parameter
+
+        Args:
+            param: LTI parameter name
+
+        Returns:
+            The value of the LTI parameter, as a str or a List, depending on the parameter.
+        """
+        if param in LAUNCH_PARAMS_IS_LIST:
+            return [x.strip() for x in self._params[param].split(",")]
+        return self._params[param]
+
+    @staticmethod
+    def valid_param(param: str) -> bool:
+        """Checks if a LTI parameter is valid or not
+
+        Args:
+            param: LTI parameter name
+
+        Returns:
+            bool: True if the parameter is valid, False otherwise.
+
+        """
+        if param.startswith("custom_") or param.startswith("ext_"):
+            return True
+        return param in LAUNCH_PARAMS
+
+    def __len__(self):
+        return len(self._params)
+
+    def __getitem__(self, item):
+        if not self.valid_param(item):
+            raise KeyError("{} is not a valid launch param".format(item))
+        try:
+            return self._param_value(item)
+        except KeyError:
+            # catch and raise new KeyError in the proper context
+            raise KeyError(item)
+
+    def __setitem__(self, key, value):
+        if not self.valid_param(key):
+            raise InvalidLaunchParamException(key)
+        if key in LAUNCH_PARAMS_IS_LIST:
+            if isinstance(value, list):
+                value = ",".join([x.strip() for x in value])
+        self._params[key] = value
+
+    def __delitem__(self, key):
+        if key in self._params:
+            del self._params[key]
+
+    def __iter__(self):
+        return iter(self._params)
+
+    @property
+    def urlencoded(self) -> str:
+        """Get the URL encoded representation of the LTI parameter list
+
+        Returns:
+            str: URL encoded LTI parameters
+        """
+        params = dict(self)
+        # stringify any list values
+        for key, value in params.items():
+            if isinstance(value, list):
+                params[key] = ",".join(value)
+        return urlencode(params)
+
+
+class LaunchParamException(Exception):
+    """Custom Exception related to LTI launch param processing"""
+
+
+class InvalidLaunchParamException(LaunchParamException):
+    """Custom Exception thrown when an invalid parameter is found in a LTI launch request"""
+
+    def __init__(self, param):
+        message = "{} is not a valid launch param".format(param)
+        super().__init__(message)
+
+
+class MissingLaunchParamException(LaunchParamException):
+    """Custom Exception thrown when a required LTI parameter is missing in an LTI launch request"""
+
+    def __init__(self, param):
+        message = "missing launch param : {}".format(param)
+        super().__init__(message)
```

Comment:
We usually define exceptions in a separated module called... `exceptions.py` :smile: